### PR TITLE
Enhance the output format for the function CallProfiler::print_statistics in the profiler

### PR DESCRIPTION
**Pull request by Vincent-Lien**
I improved the function call output format for Issue #421, transformed function call statistics from plain text to a structured table format.

To draw the divider for the table conveniently, I rounded the output numbers for Total Time (s), Per Call (s), Cumulative Time (s), and Per Call (s) to six decimal places.

The origin table show as below:
```
                                        7 function calls in 2.40007 seconds

                           Function Name               Call Count           Total Time (s)             Per Call (s)      Cumulative Time (s)             Per Call (s)
                                     bar                        4                  2.00004                 0.500009                  2.00004                 0.500009
                                     baz                        1                 0.800013                 0.800013                 0.200003                 0.200003
                                     foo                        2                  1.20004                 0.600022                 0.200027                 0.100014
```

And I transformed the table into:
```
                                                 7 function calls in 2.40002 seconds

---------------------------------------------------------------------------------------------------------------------------------------
|    Function Name    |    Call Count    |    Total Time (s)    |    Per Call (s)    |    Cumulative Time (s)    |    Per Call (s)    |
---------------------------------------------------------------------------------------------------------------------------------------
|        bar          |        4         |       2.000012       |      0.500003      |          2.000012         |      0.500003      |
|        baz          |        1         |       0.800010       |      0.800010      |          0.200006         |      0.200006      |
|        foo          |        2         |       1.200014       |      0.600007      |          0.200007         |      0.100003      |
---------------------------------------------------------------------------------------------------------------------------------------
```

### CODE REVIEW

File: cpp/modmesh/toggle/profile.cpp [hunk 1]
```diff
@@ -36,31 +36,46 @@ std::string TimeRegistry::detailed_report() const
     std::ostringstream ostm;
     /// Header
     ostm
-        << std::setw(40) << total_call_count()
+        << std::setw(50) << total_call_count()
```
Comment (Vincent-Lien) [inline cpp/modmesh/toggle/profile.cpp]: Add 10 blanks before the header to set it at the center.

File: cpp/modmesh/toggle/profile.cpp [hunk 2]
```diff
@@ -36,31 +36,46 @@ std::string TimeRegistry::detailed_report() const
...
         << " seconds" << std::endl;
     ostm
         << std::endl
-        << std::setw(40) << "Function Name"
-        << std::setw(25) << "Call Count"
-        << std::setw(25) << "Total Time (s)"
-        << std::setw(25) << "Per Call (s)"
-        << std::setw(25) << "Cumulative Time (s)"
-        << std::setw(25) << "Per Call (s)"
+        << std::setfill('-') << std::setw(135) << "";
```
Comment (Vincent-Lien) [inline cpp/modmesh/toggle/profile.cpp]: Add the top line of the table

File: cpp/modmesh/toggle/profile.cpp [hunk 3]
```diff
@@ -36,31 +36,46 @@ std::string TimeRegistry::detailed_report() const
...
+    ostm
+        << std::endl
+        << std::setw(4)
+        << "|    Function Name    "
+        << "|    Call Count    "
+        << "|    Total Time (s)    "
+        << "|    Per Call (s)    "
+        << "|    Cumulative Time (s)    "
+        << "|    Per Call (s)    |"
+        << std::endl;
```
Comment (Vincent-Lien) [inline cpp/modmesh/toggle/profile.cpp]: Add dividers between the subtitle of each columns

File: cpp/modmesh/toggle/profile.cpp [hunk 4]
```diff
@@ -36,31 +36,46 @@ std::string TimeRegistry::detailed_report() const
...
+    ostm
+        << std::setfill('-') << std::setw(135) << ""
```
Comment (Vincent-Lien) [inline cpp/modmesh/toggle/profile.cpp]: Add the divider between subtitle and numbers

File: cpp/modmesh/toggle/profile.cpp [hunk 5]
```diff
@@ -36,31 +36,46 @@ std::string TimeRegistry::detailed_report() const
...
-            << std::setw(25) << it->second.ctime()
-            << std::setw(25) << it->second.ctime() / it->second.count()
+            << std::fixed << std::setprecision(6)
+            << std::setw(4)                            << "|        " 
+            << it->first                               << "          |        "
+            << it->second.count()                      << "         |       "
+            << it->second.time()                       << "       |      "
+            << it->second.time() / it->second.count()  << "      |          "
+            << it->second.ctime()                      << "         |      "
+            << it->second.ctime() / it->second.count() << "      |"
```
Comment (Vincent-Lien) [inline cpp/modmesh/toggle/profile.cpp]: Add dividers between numbers of each columns

File: cpp/modmesh/toggle/profile.cpp [hunk 6]
```diff
@@ -36,31 +36,46 @@ std::string TimeRegistry::detailed_report() const
...
             << std::endl;
     }
+
+    ///Endline
+    ostm
+        << std::setfill('-') << std::setw(135) << ""
+        << std::endl;
+
```
Comment (Vincent-Lien) [inline cpp/modmesh/toggle/profile.cpp]: Add bottom line of the table

File: tests/test_callprofiler.py [hunk 7]
```diff
@@ -221,7 +221,7 @@ def baz():
 
         # Check the number of lines
         stat_line_list = root_stat.split("\n")
-        self.assertEqual(len(stat_line_list), 7)
+        self.assertEqual(len(stat_line_list), 10)
```
Comment (Vincent-Lien) [inline tests/test_callprofiler.py]: Because added the dividers, the lines of the table become 10 lines. I changed the lines from 7 to 10.

File: tests/test_callprofiler.py [hunk 8]
```diff
@@ -238,18 +238,31 @@ def baz():
 
         # Check the third line
         ref_line3 = (
-            "                           Function Name"
-            + "               Call Count"
-            + "           Total Time (s)"
-            + "             Per Call (s)"
-            + "      Cumulative Time (s)"
-            + "             Per Call (s)"
+            "-" * 135
```
Comment (Vincent-Lien) [inline tests/test_callprofiler.py]: The third line become the top line of the table. This line is for checking the third line of the table.

File: tests/test_callprofiler.py [hunk 9]
```diff
@@ -238,18 +238,31 @@ def baz():
...
+        # Check the forth line
+        ref_line4 = (
+              "|    Function Name    "
+            + "|    Call Count    "
+            + "|    Total Time (s)    "
+            + "|    Per Call (s)    "
+            + "|    Cumulative Time (s)    "
+            + "|    Per Call (s)    |"
+        )
+        self.assertEqual(stat_line_list[3], ref_line4)
```
Comment (Vincent-Lien) [inline tests/test_callprofiler.py]: The forth line is the subtitle in the table. This line is for checking the forth line of the table.

File: tests/test_callprofiler.py [hunk 10]
```diff
@@ -238,18 +238,31 @@ def baz():
...
+
+        # Check the fifth line
+        ref_line5 = (
+            "-" * 135
+        )
+        self.assertEqual(stat_line_list[4], ref_line5)
```
Comment (Vincent-Lien) [inline tests/test_callprofiler.py]: The fifth line is the divider between subtitles and numbers. This line is for checking the fifth line of the table.

File: tests/test_callprofiler.py [hunk 11]
```diff
@@ -238,18 +238,31 @@ def baz():
...
+
+        # Check sixth to eighth lines
         stat_dict = {}
-        for line in stat_line_list[3:-1]:
+        for line in stat_line_list[5:-2]:
+            line = line.replace('|', '')
```
Comment (Vincent-Lien) [inline tests/test_callprofiler.py]: The sixth to eighth lines are numbers of the output. I changed the index of `stat_line_list` to `[5:-2]` to meet the correct lines. Furthermore, because added the divide line `'|'`, I replace the `'|'` by `''` to sure the checking step below work correctly.

File: tests/test_callprofiler.py [hunk 12]
```diff
@@ -285,6 +298,12 @@ def baz():
         self.assertGreaterEqual(baz_dict["cumulative_time"], time3)
         self.assertGreaterEqual(baz_dict["cumulative_per_call"], time3)
 
+        # Check the ninth line
+        ref_line9 = (
+            "-" * 135
+        )
+        self.assertEqual(stat_line_list[8], ref_line9)
+
```
Comment (Vincent-Lien) [inline tests/test_callprofiler.py]: The ninth line is the bottom line of the table. This line is for checking the ninth line of the table.

**Review by Vincent-Lien — COMMENTED** [review]
I improved the function call output format.

File: cpp/modmesh/toggle/profile.cpp (hunk 1 above)
Comment (tigercosmos) [reply cpp/modmesh/toggle/profile.cpp]: Maybe you can consider making all these numbers as constant values.

File: cpp/modmesh/toggle/profile.cpp (hunk 5 above)
Comment (tigercosmos) [reply cpp/modmesh/toggle/profile.cpp]: Maybe we can have a better way to deal with `"          |        "`. This looks hard to maintain.

File: tests/test_callprofiler.py (hunk 11 above)
Comment (tigercosmos) [reply tests/test_callprofiler.py]: Consider adding comments in the code, especially for `[5:-2]`. I guess people will not understand later when they look at the code.

File: cpp/modmesh/toggle/profile.cpp (hunk 1 above)
Comment (yungyuc) [reply cpp/modmesh/toggle/profile.cpp]: `constexpr` sounds like a good idea at the moment.  The code looks like:

```cpp
constexpr int width = 50;
ostm
     << std::setw(width) << total_call_count();
```

File: cpp/modmesh/toggle/profile.cpp (hunk 5 above)
Comment (yungyuc) [reply cpp/modmesh/toggle/profile.cpp]: I don't like the table frames.  It makes parsing impossible.  Please avoid table frames at all.

**Review by yungyuc — CHANGES_REQUESTED** [review]
The frameless table format looks nice.  What are benefits of changing it?

```
                                        7 function calls in 2.40007 seconds

                           Function Name               Call Count           Total Time (s)             Per Call (s)      Cumulative Time (s)             Per Call (s)
                                     bar                        4                  2.00004                 0.500009                  2.00004                 0.500009
                                     baz                        1                 0.800013                 0.800013                 0.200003                 0.200003
                                     foo                        2                  1.20004                 0.600022                 0.200027                 0.100014
```

----

Changes requested:

- [x] Do not use table frames.

**Comment by yungyuc** [comment]
Please make sure all CI pass before request for the next review.

**Comment by ThreeMonth03** [comment]
> The frameless table format looks nice. What are benefits of changing it?
> 
> ```
>                                         7 function calls in 2.40007 seconds
> 
>                            Function Name               Call Count           Total Time (s)             Per Call (s)      Cumulative Time (s)             Per Call (s)
>                                      bar                        4                  2.00004                 0.500009                  2.00004                 0.500009
>                                      baz                        1                 0.800013                 0.800013                 0.200003                 0.200003
>                                      foo                        2                  1.20004                 0.600022                 0.200027                 0.100014
> ```
> 
> Changes requested:
> 
> * [ ]  Do not use table frames.

You are right, we cannot benefit from the frame table. However, there are some flaws in our current frameless table. The header ` 7 function calls in 2.40007 seconds` is slightly skewed to the left. Moreover, if the number of seconds or the function name is sufficiently large, the original table will overflow, and the layout would break.

**Comment by yungyuc** [comment]
> You are right, we cannot benefit from the frame table. However, there are some flaws in our current frameless table. The header ` 7 function calls in 2.40007 seconds` is slightly skewed to the left. Moreover, if the number of seconds or the function name is sufficiently large, the original table will overflow, and the layout would break.

The header alignment should be improved.  Justifying with decimal point is nice to have (not a must).

Size overflow is very hard to deal with when using a table format.  But we can live with it when the table does not have a frame.  The frame will make it look worse when the overflow happens.  We do not need to spend too much time for it for now, unless there is not an important thing to do.

File: cpp/modmesh/toggle/profile.cpp (hunk 1 above)
Comment (Vincent-Lien) [reply cpp/modmesh/toggle/profile.cpp]: The constant indeed makes the code clearer. I'll add it to the code.

File: tests/test_callprofiler.py (hunk 11 above)
Comment (Vincent-Lien) [reply tests/test_callprofiler.py]: Ok, I'll add comments to make it easier to understand.

File: cpp/modmesh/toggle/profile.cpp (hunk 5 above)
Comment (Vincent-Lien) [reply cpp/modmesh/toggle/profile.cpp]: @tigercosmos  Got it. If there are similar tasks in the future, I'll use another way to represent it. But as @yungyuc  said, the frame makes it hard to parse, so I'll remove the frame this time.

**Comment by Vincent-Lien** [comment]
> The frameless table format looks nice. What are benefits of changing it?
> 
> ```
>                                         7 function calls in 2.40007 seconds
> 
>                            Function Name               Call Count           Total Time (s)             Per Call (s)      Cumulative Time (s)             Per Call (s)
>                                      bar                        4                  2.00004                 0.500009                  2.00004                 0.500009
>                                      baz                        1                 0.800013                 0.800013                 0.200003                 0.200003
>                                      foo                        2                  1.20004                 0.600022                 0.200027                 0.100014
> ```
> 
> Changes requested:
> 
> * [ ]  Do not use table frames.

In my opinion, frames help visually separate data points and align columns, reducing the chance of misinterpretation. But it indeed makes it hard to parse, so I'll remove the table frame.

**Comment by yungyuc** [comment]
@Vincent-Lien , when you are ready for the next review, please leave a global comment.

And it's been 2 weeks since the last review.  If you intend to continue to work on this PR, could you please make an estimate for when you want to finish it?

**Comment by Vincent-Lien** [comment]
> @Vincent-Lien , when you are ready for the next review, please leave a global comment.
> 
> And it's been 2 weeks since the last review. If you intend to continue to work on this PR, could you please make an estimate for when you want to finish it?

@yungyuc According to the discussion above, it seems that creating a new branch and starting from scratch would be easier to deal with this issue. 

Furthermore, I've made some unnecessary commits over the past few weeks due to my unfamiliarity with code verification, so I'd prefer to start fresh on a new branch. 

If I were to create a new branch, should I delete the current one? What would happen to this PR? Or would you recommend continuing to work on this branch? Alternatively, I could try resetting the commits to when I created the branch and push --force, but I'm not sure if this operation is safe.

Please give me three more days, and I'll try to finish the new version by Thursday (Dec. 5th).

**Comment by yungyuc** [comment]
> @yungyuc According to the discussion above, it seems that creating a new branch and starting from scratch would be easier to deal with this issue.
> 
> Furthermore, I've made some unnecessary commits over the past few weeks due to my unfamiliarity with code verification, so I'd prefer to start fresh on a new branch.

Please keep using this branch for PR.  You can keep your current changesets by using a private aliasing branch, e.g.,

```
$ git co -b backup/current-but-nonideal-code # create you "backup" branch
$ git co -b pref/table # switch back to the PR branch
$ git squash # to get your pref/table branch clean
$ git push -f # force push to update the PR to a clean state
```

Go to discord https://discord.gg/P9U7PFv to seek help if you need.

> Please give me three more days, and I'll try to finish the new version by Thursday (Dec. 5th).

An estimate is all we need. If you cannot make it by the time you planned, just leave a comment by then and set a new schedule.

**Comment by Vincent-Lien** [comment]
I set more space to make header to the middle, and align output numbers by setting precision to 6 decimal places.
The new table will be like:
```
                                                                              7 function calls in 2.40004 seconds

                           Function Name               Call Count           Total Time (s)             Per Call (s)      Cumulative Time (s)             Per Call (s)
                                     bar                        4                 2.000023                 0.500006                 2.000023                 0.500006
                                     baz                        1                 0.800016                 0.800016                 0.200007                 0.200007
                                     foo                        2                 1.200022                 0.600011                 0.200014                 0.100007
```

**Review by yungyuc — APPROVED** [review]
> I set more space to make header to the middle, and align output numbers by setting precision to 6 decimal places. The new table will be like:
> 
> ```
>                                                                               7 function calls in 2.40004 seconds
> 
>                            Function Name               Call Count           Total Time (s)             Per Call (s)      Cumulative Time (s)             Per Call (s)
>                                      bar                        4                 2.000023                 0.500006                 2.000023                 0.500006
>                                      baz                        1                 0.800016                 0.800016                 0.200007                 0.200007
>                                      foo                        2                 1.200022                 0.600011                 0.200014                 0.100007
> ```

It is too wide, but I can live with it for now.  Thanks for the PR.
